dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.cpp

```diff
@@ -0,0 +1,265 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Executor/PipelineExecutorContext.h>
+#include <Flash/Executor/toRU.h>
+#include <Flash/Pipeline/Schedule/TaskQueues/IOPriorityQueue.h>
+#include <Flash/Pipeline/Schedule/TaskQueues/MultiLevelFeedbackQueue.h>
+#include <Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.h>
+#include <Flash/Pipeline/Schedule/Tasks/TaskHelper.h>
+
+namespace DB
+{
+template <typename NestedQueueType>
+void ResourceControlQueue<NestedQueueType>::submit(TaskPtr && task)
+{
+    std::lock_guard lock(mu);
+    submitWithoutLock(std::move(task));
+}
+
+template <typename NestedQueueType>
+void ResourceControlQueue<NestedQueueType>::submit(std::vector<TaskPtr> & tasks)
+{
+    std::lock_guard lock(mu);
+    for (auto & task : tasks)
+    {
+        submitWithoutLock(std::move(task));
+    }
+}
+
+template <typename NestedQueueType>
+void ResourceControlQueue<NestedQueueType>::submitWithoutLock(TaskPtr && task)
+{
+    if unlikely (is_finished)
+    {
+        FINALIZE_TASK(task);
+        return;
+    }
+    const auto & query_id = task->getQueryId();
+    if unlikely (cancel_query_id_cache.contains(query_id))
+    {
+        cancel_task_queue.push_back(std::move(task));
+        return;
+    }
+
+    // name can be empty, it means resource control is disabled.
+    const std::string & name = task->getResourceGroupName();
```

Comment:
Who has higher priority, "" or "default"?
If the resource control is turned on at runtime, both "" and default will appear.
Need to treat "" as "default"?

Reply:
"default" resource group will talk with GAC to fetch tokens as other resource group.
But ""(when resource control is disabled) should not.

Reply:
In the next LAC pr, priority of "" is always `1.0`, which is higher than "default".

Reply:
Does this mean that "" has the highest priority, even if the user has set a resource group with the highest priority?

Reply:
Yes

Reply:
OK, I have some doubts, does tidb/tikv handle it in the same way?